### PR TITLE
feat: add timezone support to medication schedules

### PR DESCRIPTION
**Pull request by vfilby**
## Summary

This PR adds comprehensive timezone support to medication schedules, ensuring that medication reminders and dose tracking work correctly when users travel across timezones.

**Key Changes:**
- Add `timezone` field to medication schedules (stores IANA timezone identifier)
- Implement timezone-aware date calculations for "was logged today" checks
- Fix critical bug in UTC timestamp conversion that caused incorrect dose tracking
- Add performance optimizations (formatter caching)
- Includes database migration (v19) with backward compatibility

## Problem Statement

Previously, medication schedules used device-local time, causing issues when users travel:
- Notifications would fire at incorrect times in new timezones
- "Already logged" checks would fail when device timezone differs from schedule timezone
- No way to preserve original schedule intent when traveling

### 1. Database Schema Changes
- Added `timezone` column to `medication_schedules` table
- Migration sets device timezone as default for existing schedules
- Schema validation ensures valid IANA timezone identifiers

### 2. Timezone-Aware Timestamp Conversion
Implemented `getUTCTimestampInTimezone()` function that:
- Converts local time in specific timezone to UTC timestamp
- Handles DST transitions (spring forward/fall back)
- Uses iterative algorithm for accuracy
- Includes formatter caching for performance

### 3. Updated Notification Logic
- `wasLoggedForScheduleToday()` now uses schedule's timezone, not device timezone
- Notification handlers correctly check doses in schedule's original timezone
- Prevents duplicate notifications when traveling

## Technical Details

**DST Handling:**
- Spring forward (non-existent times): Auto-adjusts to next valid time
- Fall back (ambiguous times): Uses first occurrence

**Performance:**
- Formatter instances cached per timezone (unbounded cache - acceptable for typical use)
- Typically 1-2 timezones cached per user

**Migration Safety:**
- Idempotent migration (checks if column exists)
- Rollback support (complex due to SQLite limitations)
- Default timezone based on device's current timezone

## Test Coverage

Added comprehensive tests covering:
- ✅ Basic timezone-aware checking (multiple IANA timezones)
- ✅ Timezone changes (PDT to EDT scenarios)
- ✅ UTC timestamp calculation accuracy
- ✅ DST transition edge cases (spring forward, fall back)
- ✅ Schema validation for invalid timezones
- ✅ Migration integration tests

**Test Results:**
- 191 new lines of test code in medicationRepository.test.ts
- All existing tests pass with timezone parameter added

## Known Limitations

1. **Migration Default Timezone**: Existing schedules use device's current timezone at migration time, not the timezone when schedules were originally created. Users who created schedules while traveling may need to manually update them.

2. **No User Notification**: Migration happens silently. Consider adding release notes or in-app notification.

3. **Cache Growth**: Formatter cache has no eviction policy (unbounded growth). Impact is minimal for typical users (1-2 timezones), but could accumulate in edge cases.

## Files Changed

- `medicationRepository.ts`: Core timezone conversion logic (124 lines added)
- `migrations.ts`: Database migration v19 (67 lines)
- `notificationService.ts`: Updated to use schedule timezones (17 lines changed)
- `MedicationScheduleManager.tsx`: Auto-detect device timezone for new schedules
- Schema, types, and validation updates
- Comprehensive test coverage

## Breaking Changes

⚠️ **Database Migration Required**: Users will need to run migration v19 on app update.

## Test Plan

- [x] Unit tests pass (191+ new test lines)
- [x] Migration tests pass (integration test updated to v19)
- [x] Schema validation tests pass
- [x] Notification service tests updated and passing
- [ ] Manual testing: Create schedule, travel to different timezone, verify notifications
- [ ] Manual testing: Verify existing schedules after migration

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: app/src/components/__tests__/MedicationScheduleManager.test.tsx [hunk 1]
```diff
@@ -80,6 +82,7 @@ describe('MedicationScheduleManager', () => {
       expect(mockOnSchedulesChange).toHaveBeenCalledWith([
         {
           time: '09:00',
+          timezone: expect.any(String), // Device timezone from Intl.DateTimeFormat
```
Comment (vfilby) [inline app/src/components/__tests__/MedicationScheduleManager.test.tsx]: Should this be more than just a string? We could Validate against the list, or validate that the first portion is one of the regions.  Or that it is in the format (region)/(city)?

File: app/src/database/__tests__/medicationRepository.test.ts [hunk 2]
```diff
@@ -818,6 +823,327 @@ describe('medicationDoseRepository', () => {
...
+          ).resolves.toBe(false);
+
+          expect(mockDatabase.getAllAsync).toHaveBeenCalled();
+        });
+
+        it('should handle empty timezone string', async () => {
+          const medicationId = 'med-123';
+          const scheduleId = 'schedule-123';
+          const scheduledTime = '09:00';
+          const emptyTimezone = '';
```
Comment (vfilby) [inline app/src/database/__tests__/medicationRepository.test.ts]: What does an empty TZ mean semantically?

File: app/src/database/medicationRepository.ts [hunk 3]
```diff
@@ -198,6 +198,101 @@ export const medicationRepository = {
...
+): number {
+  // Get cached formatter for the target timezone
+  const formatter = getFormatter(timezone);
+
+  // Helper to extract date components from a UTC timestamp as viewed in the target timezone
+  const getLocalComponents = (utcTimestamp: number) => {
+    const parts = formatter.formatToParts(new Date(utcTimestamp));
+    return {
+      year: parseInt(parts.find(p => p.type === 'year')!.value),
+      month: parseInt(parts.find(p => p.type === 'month')!.value) - 1,
```
Comment (vfilby) [inline app/src/database/medicationRepository.ts]: Why the '- 1'?

**Comment by vfilby** [comment]
/oc can you review the timezone logic that was added?

File: app/src/components/__tests__/MedicationScheduleManager.test.tsx (hunk 1 above)
Comment (vfilby) [reply app/src/components/__tests__/MedicationScheduleManager.test.tsx]: Fixed! Changed from `expect.any(String)` to `expect.stringMatching(/^[A-Za-z]+\/[A-Za-z_]+$/)` to validate proper IANA timezone format (Region/City).

This ensures the timezone auto-detection is producing valid IANA identifiers like 'America/Los_Angeles' rather than arbitrary strings.

File: app/src/database/medicationRepository.ts [hunk 4]
```diff
@@ -198,6 +198,101 @@ export const medicationRepository = {
...
+
+/**
+ * Get a cached or create a new Intl.DateTimeFormat for the given timezone.
+ * @param timezone IANA timezone identifier
+ * @returns Cached DateTimeFormat instance
+ */
+function getFormatter(timezone: string): Intl.DateTimeFormat {
+  if (!formatterCache.has(timezone)) {
+    formatterCache.set(timezone, new Intl.DateTimeFormat('en-US', {
+      timeZone: timezone,
```
[the remark this answers is not in the document]
Comment (vfilby) [reply app/src/database/medicationRepository.ts]: Same as above - empty strings throw an error in `Intl.DateTimeFormat` validation and fallback to device timezone.

Added explicit comment in the code:
```typescript
// Empty strings, null, undefined, and malformed timezone IDs all throw errors from Intl.DateTimeFormat
let validatedTimezone = scheduleTimezone;
try {
  Intl.DateTimeFormat(undefined, { timeZone: scheduleTimezone });
} catch (error) {
  logger.error(`Invalid timezone: ${scheduleTimezone}, falling back to device timezone`, error);
  validatedTimezone = Intl.DateTimeFormat().resolvedOptions().timeZone;
}
```

The try/catch handles all invalid cases including empty strings.

File: app/src/database/medicationRepository.ts [hunk 5]
```diff
@@ -198,6 +198,101 @@ export const medicationRepository = {
...
+const formatterCache = new Map<string, Intl.DateTimeFormat>();
+
+/**
+ * Get a cached or create a new Intl.DateTimeFormat for the given timezone.
+ * @param timezone IANA timezone identifier
+ * @returns Cached DateTimeFormat instance
+ */
+function getFormatter(timezone: string): Intl.DateTimeFormat {
+  if (!formatterCache.has(timezone)) {
+    formatterCache.set(timezone, new Intl.DateTimeFormat('en-US', {
```
[the remark this answers is not in the document]
Comment (vfilby) [reply app/src/database/medicationRepository.ts]: Good point! I've addressed this by:

1. **Added bounded cache**: Set `MAX_CACHE_SIZE = 10` with FIFO eviction to prevent unbounded growth
2. **Enhanced documentation** explaining why caching is beneficial:

```typescript
/**
 * Cache for Intl.DateTimeFormat instances to improve performance within the iterative
 * UTC conversion algorithm. Each timestamp calculation may call formatToParts() up to
 * 3 times to handle DST transitions, making formatter reuse beneficial.
 *
 * While single-user mobile apps typically use 1-2 timezones, we limit cache size to
 * prevent unbounded growth in edge cases (e.g., frequent international travel).
 * Size limit of 10 accommodates reasonable travel scenarios without memory concerns.
 */
```

The caching is primarily for the **iterative algorithm** which calls `formatToParts()` up to 3 times per timestamp conversion. So even a single-timezone user benefits from not recreating the formatter on each iteration.

The 10-entry limit handles edge cases (frequent travelers) while keeping memory footprint minimal.

**Comment by vfilby** [comment]
## ✅ CI Tests Now Passing

Fixed the test failures! The issue was that the timezone validation regex only matched `Region/City` format (like `America/Los_Angeles`) but the CI environment uses `UTC` which doesn't have a slash.

**Updated regex**: `/^[A-Za-z]+([/_][A-Za-z_]+)?$/`

Now accepts both:
- Simple formats: `UTC`, `GMT`, `EST`, `PST`
- Region/City formats: `America/Los_Angeles`, `Europe/London`

All checks are now passing ✅
